torch/csrc/jit/script/module.h

```diff
@@ -381,120 +393,291 @@ struct TORCH_API Module {
-  Module module;
-};
+namespace detail {
 
-struct NameValue {
-  std::string name;
-  IValue value;
+struct TORCH_API SlotCursor {
+  Module module_;
+  int64_t i_; // slot offset, -1 indicates the module itself
 };
 
-// this iterator for the slot list defined below has a position in the list i_
-// and an optional field type_ that if present
-// restricts iteration to only the slots of module_ that
-// have EntityType *type_. This allows it to return, e.g.
-// only the parameter slots.
-// The template parameter allows us to use the same implementation for a list
-// that returns Module via template specialization of the operator* method.
-template <typename T>
+} // namespace detail
+
+// This iterator allows the (optionally recursive) enumeration of
+// the  members of a Module. It performs a depth-first pre-order
+// traversal of the module. The Policy template parameter determines
+// which slots of the object should be include. For instance,
+// when iterating parameters, we return the parameter tensors,
+// but skip modules, buffers, and other attributes.
+// See ModulePolicy for comments about Policy object's API.
+template <typename Policy>
 struct TORCH_API slot_iterator_impl {
-  slot_iterator_impl(Module module, c10::optional<EntityType> type, size_t i)
-      : module_(module), type_(type), i_(i) {
-    advance_to_valid();
-  }
-  T operator*() const;
-  T operator->() const {
+  using SlotCursor = detail::SlotCursor;
+  using value_type = typename Policy::value_type;
+  slot_iterator_impl(
+      Module root,
+      bool recurse, // if true, do a depth-first search, otherwise, just look at
+                    // slots of root
+      bool return_module) // if true include root itself as the first thing
+                          // visited (used in modules())
+      : cursors_({SlotCursor{root, return_module ? -1 : 0}}),
+        recurse_(recurse) {
+    // advance iterator to first valid element (or the end, if empty)
+    while_not_valid_next();
+  }
+  // empty frame_, represents end of iteration
+  slot_iterator_impl() : recurse_(false) {}
+  value_type operator*() const {
+    return Policy::create(cursors_, cur());
+  }
+  value_type operator->() const {
     return **this;
   }
   slot_iterator_impl& operator++() {
-    ++i_;
-    advance_to_valid();
+    next_valid();
     return *this;
   }
   slot_iterator_impl operator++(int) {
+    // this is really expensive, should we delete it so people don't use it
+    // instead of prefix?
     slot_iterator_impl old = *this;
     ++(*this);
     return old;
   }
 
  private:
-  void advance_to_valid() {
-    while (i_ < module_.num_slots() &&
-           (type_ && module_.entity_type(i_) != *type_)) {
-      ++i_;
+  // return_module() is a corner case where instead of returning a submodule
+  // of root, we are return root itself, because we are iterating modules(),
+  // which contains the root module itself.
+  // It is represented with a single Frame whose index is -1.
```

Comment:
Nit: since Frame was renamed to SlotCursor, we need to update all its references accordingly.